libraries/chain/include/eosio/chain/wasm_interface_private.hpp

```diff
@@ -84,48 +124,90 @@ namespace eosio { namespace chain {
+         // Safe to update the cache without locking.
+         // Anything last used before or on the LIB can be evicted.
          const auto first_it = wasm_instantiation_cache.get<by_last_block_num>().begin();
          const auto last_it  = wasm_instantiation_cache.get<by_last_block_num>().upper_bound(lib);
-         if (callback) {
-            for(auto it = first_it; it != last_it; it++) {
-               callback(it->code_hash, it->vm_version);
-            }
-         }
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+         if(eosvmoc) for(auto it = first_it; it != last_it; it++)
+            eosvmoc->cc.free_code(it->code_hash, it->vm_version);
+#endif
          wasm_instantiation_cache.get<by_last_block_num>().erase(first_it, last_it);
       }
 
-      const std::unique_ptr<wasm_instantiated_module_interface>& get_instantiated_module( const digest_type& code_hash, const uint8_t& vm_type,
-                                                                                 const uint8_t& vm_version, transaction_context& trx_context )
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+      bool is_eos_vm_oc_enabled() const {
+         return (eosvmoc || wasm_runtime_time == wasm_interface::vm_type::eos_vm_oc);
+      }
+#endif
+
+      const std::unique_ptr<wasm_instantiated_module_interface>& get_instantiated_module(
+         const digest_type&   code_hash,
+         const uint8_t&       vm_type,
+         const uint8_t&       vm_version,
+         transaction_context& trx_context)
       {
-         wasm_cache_index::iterator it = wasm_instantiation_cache.find(
-                                             boost::make_tuple(code_hash, vm_type, vm_version) );
-         const code_object* codeobject = nullptr;
-         if(it == wasm_instantiation_cache.end()) {
-            codeobject = &db.get<code_object,by_code_hash>(boost::make_tuple(code_hash, vm_type, vm_version));
-
-            it = wasm_instantiation_cache.emplace( wasm_interface_impl::wasm_cache_entry{
-                                                      .code_hash = code_hash,
-                                                      .last_block_num_used = UINT32_MAX,
-                                                      .module = nullptr,
-                                                      .vm_type = vm_type,
-                                                      .vm_version = vm_version
-                                                   } ).first;
+         wasm_cache_index::iterator it;
+         if (trx_context.control.is_write_window()) {
+            // When in write window (either read only threads are not enabled or
+            // they are not schedued to run), only main thread is processing
+            // transactions. No need to lock.
+            it = wasm_instantiation_cache.find( boost::make_tuple(code_hash, vm_type, vm_version) );
+         } else {
+            std::shared_lock g(instantiation_cache_mutex);
```

Comment:
I wonder if this is actually better than just acquiring a `scope_lock` on a non-shared mutex. It would simplify the code to just use a normal lock and not acquire it twice when you have to build and add to the cache.

Reply:
Since an instantiated module is likely in the cache, `find` (read operation)  succeed most of the time. That's why I chose shared mutex. But that made code complicated. Prefer to a simpler code. Thanks.